features/lorawan/LoRaWANInterface.cpp

```diff
@@ -0,0 +1,177 @@
+/**
+ * @file
+ *
+ * @brief      Implementation of LoRaWANBase
+ *
+ * Copyright (c) 2017, Arm Limited and affiliates.
+ * SPDX-License-Identifier: Apache-2.0
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "lorawan/LoRaWANInterface.h"
+
+inline LoRaWANStack& stk_obj()
+{
+    return LoRaWANStack::get_lorawan_stack();
+}
+
+LoRaWANInterface::LoRaWANInterface(LoRaRadio& radio)
+{
+    // Pass mac_handlers to radio to the radio driver after
+    // binding radio driver to PHY layer
+    radio_events_t *events = stk_obj().bind_radio_driver(radio);
+    radio.lock();
+    radio.init_radio(events);
+    radio.unlock();
+}
+
+LoRaWANInterface::~LoRaWANInterface()
+{
+}
+
+lora_mac_status_t LoRaWANInterface::initialize()
+{
+    return stk_obj().initialize_mac_layer();
+}
+
+lora_mac_status_t LoRaWANInterface::connect()
+{
+    // connection attempt without parameters.
+    // System tries to look for configuration in mbed_lib.json that can be
+    // overridden by mbed_app.json. However, if none of the json files are
+    // available (highly unlikely), we still fallback to some default parameters.
+    // Check lorawan_data_structure for fallback defaults.
+
+    lorawan_connect_t connection_params;
+
+    if (OVER_THE_AIR_ACTIVATION != 0) {
+        static uint8_t dev_eui[] = LORAWAN_DEVICE_EUI;
```

Comment:
Should this connect routine be made thread safe in case multiple threads call the function simultaneously?

Reply:
At this moment the system underneath LoRaWANInterface is NonCopyable. It's documented above the constructor of LoRaWANInterface. The intention is to use one and only object of the stack from whichever context it's being accessed